qiskit/algorithms/eigensolvers/numpy_eigensolver.py

```diff
@@ -110,68 +108,50 @@ def _check_set_k(self, operator: BaseOperator | PauliSumOp) -> None:
                 self._k = self._in_k
 
     def _solve(self, operator: BaseOperator | PauliSumOp) -> None:
-        if isinstance(operator, PauliSumOp):
-            sp_mat = operator.to_spmatrix()
+        if isinstance(operator, Operator):
```

Comment:
This is a more general comment. `BaseOperator | PauliSumOp` was chosen to be the same as that which Estimator supports. However from what I can see BaseOperator provides no standard way to access the data comprising the operator as needed here. An Operator has data, and Pauli types seems to have to_matrix(). ScalarOp is also a BaseOperator and while it has to_matrix does not have a sparse attribute hence would fail if used here. Looking at Estimator when it does an expectation, using StateVector, it may end up creating an Operator, depending on type, where its constructor inspects the object being passed for various aspects to build it out - I think ScalarOp would pass through this. I guess I am wondering more about the if its not an Operator and then its not PauliSumOp we are in the realm of all other BaseOperators that are not the Operator subclass thereof. The to_matrix(sparse=True) looks like for the Pauli based ones like PauliTable, SparsePauliOp that this is implemented in these - but again this is a specific set of these. This logic certainly now captures the Pauli types like SparsePauliOp of BaseOperator, which it did not before. Just wondering whether we can/should do better - maybe at least wrap the to_matrix(sparse=Try) in a try/except and raise perhaps an error if that can't be called that the operator type is not supported?

Reply:
It looks like a try/except is the best option, as there are several operator classes that do not subclass `BasePauli` but do allow `sparse=True`. It will also give a more appropriate error when a parameterized operator is passed in. From what I can see all `BaseOperator` classes, other than `Operator`, do expose a `to_matrix` method. Would it be worth adding this method to `Operator` and just returning `data`? `Clifford` and `CNOTDihedral` already effectively do this via `to_operator`. Could we perhaps also make `to_matrix` a required method of `BaseOperator`.

Reply:
Since primitives - especially the remote ones, I would think, need access to the data to be able to do the computation the fact that BaseOperator is the designated type but provides no defined mechanisms, indeed seems problematic. I am tagging @ikkoham and @t-imamichi to bring them into the conversation more generally. The reference Estimator relies on StateVector being able to compute an estimation of the operator (it and the operators being part of quantum_info) - for other primitives, that may use an aspect of it, like here, it can be different.

Here specifically, in the near term, I think a try/except to be able to raise a more meaningful error. Longer term it seems to me that whatever the type we indicate should have suitable method(s) for the purposes we have/envision. Whether BaseOperator (or something else) is updated or perhaps new mixin/Protocol(s) added seems like it should be the subject of a separate issue/PR to discuss and hopefully address these wider needs and revisit this try/catch later as warrants.